api/python/t4/packages.py

```diff
@@ -977,14 +977,19 @@ def filter(self, f, include_directories=False):
         Returns: list
             A list of truthy (logical key, entry) tuples.
         """
+        p = Package()
+
+        excluded_dirs = set()
         if include_directories:
             for lk, _ in self._walk_dir_meta():
-                if f(lk, self[lk.rstrip("/")]):
-                    yield (lk, self[lk.rstrip("/")])
+                if not f(lk, self[lk.rstrip("/")]):
+                    excluded_dirs = excluded_dirs.union({lk})
 
         for lk, entity in self.walk():
-            if f(lk, entity):
-                yield (lk, entity)
+            if f(lk, entity) and not any(p in excluded_dirs for p in pathlib.Path(lk).parents):
```

Comment:
I think you need a `PurePosixPath`, since these are logical keys

Reply:
Also... I think you should check the `excluded_dirs` first, then call `f`. It's probably reasonable for the user to assume `f` will never be called on keys in directories that got excluded - so it may raise an exception, etc.

Reply:
Great point! Done.